crates/ruma-client-api/src/error.rs

```diff
@@ -504,6 +524,76 @@ impl TryFrom<&AuthenticateError> for http::HeaderValue {
     }
 }
 
+/// How long a client should wait before they can try again.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[allow(clippy::exhaustive_enums)]
+pub enum RetryAfter {
+    /// The client should wait for the given duration.
+    ///
+    /// This variant should be preferred for backwards compatibility, as it will also populate the
+    /// `retry_after_ms` field in the body of the response.
+    Delay(Duration),
+    /// The client should wait for the date and time corresponding to the given timestamp.
+    DateTime(SecondsSinceUnixEpoch),
+}
+
+impl RetryAfter {
+    fn from_header_value(value: &http::HeaderValue) -> Option<Self> {
+        let bytes = value.as_bytes();
+
+        if bytes.iter().all(|b| b.is_ascii_digit()) {
+            // It should be a duration.
+            Some(Self::Delay(Duration::from_secs(u64::from_str(value.to_str().ok()?).ok()?)))
+        } else {
+            // It should be a date.
+            Some(Self::DateTime(SecondsSinceUnixEpoch(
+                UInt::try_from(date_header::parse(bytes).ok()?).ok()?,
+            )))
+        }
+    }
+}
+
+impl TryFrom<&RetryAfter> for http::HeaderValue {
+    type Error = RetryAfterIntoHttpHeaderError;
+
+    fn try_from(value: &RetryAfter) -> Result<Self, Self::Error> {
+        match value {
+            RetryAfter::Delay(duration) => Ok(duration.as_secs().into()),
+            RetryAfter::DateTime(ts) => {
+                let mut buffer = [0; 29];
+                date_header::format(ts.0.into(), &mut buffer)
+                    .map_err(|_| RetryAfterIntoHttpHeaderError::InvalidDateTime)?;
+                Ok(http::HeaderValue::from_bytes(&buffer)?)
+            }
+        }
+    }
+}
+
+/// An error when converting a [`RetryAfter`] to a [`http::HeaderValue`].
+#[derive(Debug, Error)]
+#[allow(clippy::exhaustive_enums)]
+pub enum RetryAfterIntoHttpHeaderError {
+    /// The date and time are too much in the future (the year is bigger than 9999).
+    #[error(
+        "Retry-After header serialization failed: the year of the datetime is bigger than 9999"
+    )]
+    InvalidDateTime,
+    /// The header value is invalid (should not happen).
+    #[error("Retry-After header serialization failed: {0}")]
+    InvalidHeader(#[from] http::header::InvalidHeaderValue),
```

Comment:
I think this not only _should not_ happen, it _cannot happen_ unless the `date_header` does something super weird. Seems worth using `expect` instead of having an error variant for.

Reply:
Now this error is `expect`ed.